walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,112 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requiredAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    // Returns true if an event contains at least one required or optional attendee
+    Predicate<Event> containsAllAttendees = e -> {
+      boolean containsRequiredAttendees = 
+          !Collections.disjoint(e.getAttendees(), requiredAttendees); 
+      boolean containsOptionalAttendees = 
+          !Collections.disjoint(e.getAttendees(), optionalAttendees);
+
+      return containsRequiredAttendees || containsOptionalAttendees;
+    };
+
+    Collection<Event> requiredAndOptionalEvents = 
+        filterEvents(containsAllAttendees, events);
+    Collection<TimePoint> requiredAndOptionalPoints =
+        createSortedTimePoints(requiredAndOptionalEvents); 
+    Collection<TimeRange> allAvailableTimes = 
+        getAvailableTimes(requiredAndOptionalPoints, request.getDuration());
+
+    // Return result if optional attendees can be accomodated or there is only one type of attendee
+    if (!allAvailableTimes.isEmpty() || requiredAttendees.isEmpty() 
+        || optionalAttendees.isEmpty()) {
+        return allAvailableTimes;
+    }
+
+    // Returns true if an event contains at least one required attendee
+    Predicate<Event> containsRequiredAttendees = e -> {   
+      return !Collections.disjoint(e.getAttendees(), requiredAttendees); 
+    };
+    
+    // If optional attendees cannot be accomodated, try again with only required attendees
```

Comment:
Is there no need to find the range that maximizes the number of optional attendees?

Reply:
That is the optional next step. I wanted to make a pull request for the first two required tasks before I moved on to that.